src/Compilers/CSharp/Test/Symbol/Symbols/RequiredMembersTests.cs

```diff
@@ -0,0 +1,1181 @@
+            //         private required int PrivateField;
+            Diagnostic(ErrorCode.ERR_RequiredMembersCannotBeLessVisibleThanContainingType, "PrivateField").WithArguments("Outer.PrivateProtectedClass.PrivateField", "Outer.PrivateProtectedClass").WithLocation(79, 30)
+        );
+    }
+
+    [Fact]
+    public void UsingRequiredMemberAttributeExplicitly()
+    {
+        var comp = CreateCompilationWithRequiredMembers(@"
+using System.Runtime.CompilerServices;
+[RequiredMember]
+class C
+{
+    [RequiredMember]
+    public int Prop { get; set; }
+    [RequiredMember]
+    public int Field;
+}
+");
+
+        comp.VerifyDiagnostics(
+            // (3,2): error CS9504: Do not use 'System.Runtime.CompilerServices.RequiredMemberAttribute'. Use the 'required' keyword on required fields and properties instead.
+            // [RequiredMember]
+            Diagnostic(ErrorCode.ERR_ExplicitRequiredMembers, "RequiredMember").WithLocation(3, 2),
+            // (6,6): error CS9504: Do not use 'System.Runtime.CompilerServices.RequiredMemberAttribute'. Use the 'required' keyword on required fields and properties instead.
+            //     [RequiredMember]
+            Diagnostic(ErrorCode.ERR_ExplicitRequiredMembers, "RequiredMember").WithLocation(6, 6),
+            // (8,6): error CS9504: Do not use 'System.Runtime.CompilerServices.RequiredMemberAttribute'. Use the 'required' keyword on required fields and properties instead.
+            //     [RequiredMember]
+            Diagnostic(ErrorCode.ERR_ExplicitRequiredMembers, "RequiredMember").WithLocation(8, 6),
+            // (9,16): warning CS0649: Field 'C.Field' is never assigned to, and will always have its default value 0
+            //     public int Field;
+            Diagnostic(ErrorCode.WRN_UnassignedInternalField, "Field").WithArguments("C.Field", "0").WithLocation(9, 16)
+        );
+
+        var prop = comp.SourceModule.GlobalNamespace.GetMember<PropertySymbol>("C.Prop");
+        Assert.False(prop.IsRequired);
+    }
+
+    [Fact]
+    public void UsingRequiredMemberAttributeExplicitly_WrongLocations()
+    {
+        var comp = CreateCompilation(@"
+using System;
+using System.Runtime.CompilerServices;
+class C
+{
+    [RequiredMember]
+    void M() {}
+    [RequiredMember]
+    event Action E;
+    [RequiredMember]
+    C() {}
+    [RequiredMember]
+    ~C() {}
+    [return: RequiredMember]
+    void M<[RequiredMember] T>([RequiredMember] int i) {}
+}
+
+namespace System.Runtime.CompilerServices
+{
+    public class RequiredMemberAttribute : Attribute
+    {
+        public RequiredMemberAttribute()
+        {
+        }
+    }
+}
+");
+
+        comp.VerifyDiagnostics(
+            // (9,18): warning CS0067: The event 'C.E' is never used
+            //     event Action E;
+            Diagnostic(ErrorCode.WRN_UnreferencedEvent, "E").WithArguments("C.E").WithLocation(9, 18)
+        );
+    }
+}
```

Comment:
Let's add test for `required` + initializer. Feels like that should be disallowed.
`public required int Prop { get; set; } = 42;`

Reply:
> required + initializer

I don't see a need to disallow an initializer on a required property or field. FWIW, the current design for structs makes it so the field *must* be assigned in the constructor, even if it is required (i.e. must be assigned again by the caller of the constructor).

I could see a code-quality rule having a setting to give some diagnostic to the effect of *why are you setting this if the caller needs to set it*. But I could also imagine constructors assigning some 'intermediate' value that needs to be there for construction to work, and then the user needs to assign a 'final' value later.

Reply:
1. Not sure why this would be disallowed? We could add a warning, I guess, but I don't see a reason to disallow it.
2. Likely something we should confirm with LDM.
3. We probably want to error on that, as well as properties missing setters. I'll add a prototype comment, and we'll want to follow up with LDM.

Reply:
Regarding 1: I'd be okay with a warning (fine to confirm with LDM first, not blocking PR). I just can't imagine a legitimate reason for a user to put an initializer on a required property.

Reply:
> I just can't imagine a legitimate reason for a user to put an initializer on a required property.

These are things I view as "just because we lack imagination doesn't mean some user won't find a reason". Such as a default for a `SetsRequiredMembers` constructor.